opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -372,38 +372,39 @@ def set_attribute(self, key: str, value: types.AttributeValue) -> None:
             logger.warning("invalid key (empty or null)")
             return
 
-        if isinstance(value, Sequence):
-            error_message = self._check_attribute_value_sequence(value)
-            if error_message is not None:
-                logger.warning("%s in attribute value sequence", error_message)
-                return
-            # Freeze mutable sequences defensively
-            if isinstance(value, MutableSequence):
-                value = tuple(value)
-        elif not isinstance(value, (bool, str, int, float)):
-            logger.warning("invalid type for attribute value")
+        error_message = self._check_attribute_value(value)
+        if error_message is not None:
+            logger.warning(error_message)
             return
+        # Freeze mutable sequences defensively
+        if isinstance(value, MutableSequence):
+            value = tuple(value)
 
         self.attributes[key] = value
 
     @staticmethod
-    def _check_attribute_value_sequence(sequence: Sequence) -> Optional[str]:
+    def _check_attribute_value(value: types.AttributeValue) -> Optional[str]:
         """
         Checks if sequence items are valid and are of the same type
         """
-        if len(sequence) == 0:
-            return None
+        if isinstance(value, Sequence):
+            if len(value) == 0:
+                return None
 
-        first_element_type = type(sequence[0])
+            first_element_type = type(value[0])
 
-        if first_element_type not in (bool, str, int, float):
-            return "invalid type"
+            if first_element_type not in (bool, str, int, float):
+                return "invalid type in attribute value sequence"
 
-        for element in sequence:
-            if not isinstance(element, first_element_type):
-                return "different type"
+            for element in value:
+                if not isinstance(element, first_element_type):
+                    return "different types in attribute value sequence"
+            return None
+        elif not isinstance(value, (bool, str, int, float)):
+            return "invalid type for attribute value"
```

Comment:
Similar comment here for improving the error message: if we say what was expected and what was found, that will help people debug when they see the error.